class-instant-articles-amp-markup.php

```diff
@@ -0,0 +1,163 @@
+<?php
+/**
+ * Facebook Instant Articles for WP.
+ * This source code is licensed under the license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @package default
+ * @since 4.0
+ */
+
+use Facebook\InstantArticles\AMP\AMPArticle;
+use Facebook\InstantArticles\Validators\Type;
+
+class Instant_Articles_Amp_Markup {
+
+	const SETTING_AMP_MARKUP = 'amp_markup'; /// Setting to check if AMP Markup generation is enabled
+	const SETTING_STYLE      = 'amp_stylesheet'; /// Setting that stores the JSON stylesheet
+	const SETTING_DL_MEDIA   = 'amp_download_media'; /// Enable or disable media download option
+	const QUERY_ARG          = 'amp_markup'; /// Query argument that will trigger the AMP markup generation
+
+	// To memoize the settings
+	static $settings = null;
+
+	/**
+	 * Gets the settings
+	 *
+	 * @return array The settings, check Instant_Articles_Option::get_option_decoded()
+	 * @since 4.0
+	 */
+	static function get_settings() {
+		if (self::$settings === null)
+			self::$settings = Instant_Articles_Option_Amp::get_option_decoded();
+
+		return self::$settings;
+	 }
+	/**
+	 * Checks if the AMP markup is enabled
+	 *
+	 * @return bool true if markup is enabled
+	 * @since 4.0
+	 */
+	static function is_markup_enabled() {
+		$settings = self::get_settings();
+
+		return
+			isset( $settings[self::SETTING_AMP_MARKUP] )
+			? (bool) $settings[self::SETTING_AMP_MARKUP]
+			: false;
+	}
+	/**
+	 * Adds the meta tag for AMP Markup if enabled
+	 *
+	 * @since 4.0
+	 */
+	static function inject_link_rel() {
+
+		if (!self::is_markup_enabled())
+			return;
+
+		// Transform the post to an Instant Article.
+		$adapter = new Instant_Articles_Post( get_post() );
+
+		if (!$adapter->should_submit_post())
+			return;
+
+		$url = $adapter->get_canonical_url();
+		$url = add_query_arg(self::QUERY_ARG, '1', $url );
+
+		echo '<link rel="amphtml" href="' . $url . '">';
+
+	}
+
+	/**
+	 * Generates the AMP markup if post has amp_markup
+	 *
+	 * NOTE: side-effect: function calls die() in the end.
+	 * @since 4.0
+	 */
+	static function markup_version( ) {
```

Comment:
Make it available to create filter and embed the filter generation into a separated action, this way we can allow extension for this great block.